### PR TITLE
evmzero:  Add check for range in blockHash op

### PR DESCRIPTION
**Pull request by facuMH**
This PR fixes the BLOCKHASH operation by adding a check that the number of the block to be hashed is within 256 of the current block.
This PR adds a test for the case when the block to be hashed is outside of this range, in this case the returned hash is 0 and a regression test case for the CT.

### CODE REVIEW

File: cpp/vm/evmzero/interpreter.cc [hunk 1]
```diff
@@ -865,7 +865,18 @@ struct Impl<OpCode::BLOCKHASH> {
 
   static OpResult Run(uint256_t* top, Context& ctx) noexcept {
     int64_t number = static_cast<int64_t>(top[0]);
```
Comment (HerbertJordan) [inline cpp/vm/evmzero/interpreter.cc:867]: This function does not check for overflows, but I guess it should.

File: cpp/vm/evmzero/interpreter_test.cc [hunk 2]
```diff
@@ -2845,12 +2848,29 @@ TEST(InterpreterTest, BLOCKHASH) {
...
+      .code = {op::BLOCKHASH},
+      .state_after = RunState::kDone,
+      .gas_before = 40,
+      .gas_after = 20,
+      .stack_before = {500},
+      .stack_after = {0},
+      .host = &host,
+  });
+}
+
```
Comment (HerbertJordan) [inline cpp/vm/evmzero/interpreter_test.cc:2873]: Please also add test cases for edge cases where the provided number is
- equal to the current block - {0, 1, 255, 256, 257}
- an overflow that would result in a valid value if truncated

**Review by HerbertJordan — CHANGES_REQUESTED** [review]

File: cpp/vm/evmzero/interpreter.cc [hunk 3]
```diff
@@ -864,8 +864,24 @@ struct Impl<OpCode::BLOCKHASH> {
...
+
     int64_t number = static_cast<int64_t>(top[0]);
-    top[0] = ToUint256(ctx.host->get_block_hash(number));
+    int64_t upper, lower;
+    upper = int64_t(ctx.host->get_tx_context().block_number);
+    if (upper < 257) {
+      lower = 0;
+    } else {
+      lower = upper - 256;
+    }
```
Comment (simonlechner) [inline cpp/vm/evmzero/interpreter.cc:879]: I think the reason why these checks are necessary deserves a comment.

**Review by HerbertJordan — APPROVED** [review]
Thanks for fixing this.

File: cpp/vm/evmzero/interpreter_test.cc [hunk 4]
```diff
@@ -2845,12 +2849,87 @@ TEST(InterpreterTest, BLOCKHASH) {
...
+TEST(InterpreterTest, BLOCKHASH_overflow) {
+  // const uint64_t block_number = 1000;
+  MockHost host;
+  // EXPECT_CALL(host, get_tx_context())  //
+  //     .Times(1)
+  //     .WillOnce(Return(evmc_tx_context{.block_number = block_number}));
```
Comment (simonlechner) [inline cpp/vm/evmzero/interpreter_test.cc:2920]: Should this commented code be removed?

**Review by LuisPH3 — APPROVED** [review]

**Review by HerbertJordan — APPROVED** [review]
Looks good, thanks.
